File: ros2_ws/src/py_pubsub/py_pubsub/subscriber_member_function.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String

# Create a subclass of Node for our subscriber
class MinimalSubscriber(Node):
	
	# Create the constructor for our class
	def __init__(self):
		super().__init__('minimal_subscriber')
		self.subscription = self.create_subscription( # Member function of Node
			String,   # Message type must match what the publisher is publishing
			'topic',  # We want to subscribe to the correct topic to communicate
			self.listener_callback, # Callback executes when msg is received
			10)			# Buffer for speed deficiency handling 
		self.subscription # Prevents unused variable warning
		
	# Create the listener callback
	def listener_callback(self, msg):
		# Just print the message that was received
		self.get_logger().info('I heard: "%s"' % msg.data)
		
# Create the main function 
def main(args=None):
	# Initialize rclpy
	rclpy.init(args=args)
	# Create instance of our subscriber class
	minimal_subscriber = MinimalSubscriber()
	# 'Spin' the node to get the callback going
	rclpy.spin(minimal_subscriber)
	
	# Destroy when done
	minimal_subscriber.destroy_node()
	rclpy.shutdown()



if __name__ == '__main__':
	main()
		

File: ros2_ws/src/cpp_pubsub/src/subscriber_member_function.cpp
#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "tutorial_interfaces/msg/num.hpp"

using std::placeholders::_1; 

class MinimalSubscriber : public rclcpp::Node
{
public:
	// Set the class constructor for subclass of Node
	MinimalSubscriber()
	: Node("minimal_subscriber")
	{
		// Just set the one member variable
		// Subscribe to topic, bind a callback, 
		subscription_ = this->create_subscription<tutorial_interfaces::msg::Num>(
		  "topic", 10, std::bind(&MinimalSubscriber::topic_callback, this, _1));
	}

private:
	// Create the callback function on message reception
	void topic_callback(const tutorial_interfaces::msg::Num & msg) const
	{
		// Just log the received message to the console
		RCLCPP_INFO_STREAM(this->get_logger(), "I heard: '" << msg.num << "'");
	}
	// Initialize the the member variable for a subscription
	rclcpp::Subscription<tutorial_interfaces::msg::Num>::SharedPtr subscription_;
};
	
// Create a main to run the node 
int main(int argc, char * argv[])
{
	rclcpp::init(argc,argv); // Init ros2 with args
	rclcpp::spin(std::make_shared<MinimalSubscriber>()); // Spin the node 
	rclcpp::shutdown();
	return 0;
}
